### ISSUE TITLE
basicPubSub.py hangs forever at TLSv1.2 Mutual Authentication on RPi3

### ISSUE THREAD

**Issue by kadin2048** [aws/aws-iot-device-sdk-python#25]
I'm trying to get AWSIoTPythonSDK 1.1.0 working on a Raspberry Pi 3 running Raspbian Jessie.  It _appears_ to have compatible versions of Python and OpenSSL installed, but I am getting an infinite hang on "TLSv1.2 Mutual Authentication" when I try to run basicPubSub.py.  

It doesn't give much else in the way of output and I'd be very interested in suggestions to try and get it working.

I have created the public/private keypair and downloaded the root cert as described in the docs, and moved those to the RPi, along with the latest (commit 41ead949149) SDK.  I am running Python 2.7.9 and ssl.OPENSSL_VERSION [reports](https://gist.github.com/kadin2048/4067ca04e8a1f9080de05391d6da25f9#file-versions-txt) `OpenSSL 1.0.1t  3 May 2016`, which ought to work per the README.

The command I'm running is (where MYENDPOINT and MYCERT are redactions, but represent actual values on my system):
```
python /etc/aws/python-sdk/samples/basicPubSub/basicPubSub.py  \
 -e MYENDPOINT.iot.us-east-1.amazonaws.com \   
 -r /etc/aws/deviceSDK/certs/rootCA.pem.crt \   
 -c /etc/aws/deviceSDK/certs/MYCERT-certificate.pem.crt \   
 -k /etc/aws/deviceSDK/certs/MYCERT-private.pem.key
```

I've checked all 3 file paths to make sure they're accessible.  I have permissions set to 600 on the private key, in case that was causing OpenSSL to choke (didn't seem to matter).  I've tried running as the usual "rpi" user and as root, just in case.  The output never seems to vary, and can be viewed in [this gist](https://gist.github.com/kadin2048/4067ca04e8a1f9080de05391d6da25f9#file-basicpubsub-log); it will hang with that showing until I cause a KeyboardInterrupt. 

I've verified that using the same keyfiles and endpoint, I can use the Embedded C SDK to publish and subscribe from the same device, so I know that it is definitely not a network issue.  The problem seems to be definitely related to the Python SDK.

Any suggestions of what to try?  I would really like to use the Python SDK.

**Comment by liuszeng** [comment]
Hi @kadin2048 ,

Thank you very much for your interest in AWS IoT Python SDK!

I tried to duplicate the issue on my side but got no luck. I was able to use X.509 certificate based mutual auth. to run the pubsub sample. The environment information is listed below:

    Rpi 2 Raspbian Jessie
    Python 2.7.9
    OpenSSL v1.0.1k 8 Jan 2015
    Policy attached:
    {
        ...
        "Statement":[
            "Action": "iot:*",
            "Resource": "*",
            "Effect": "Allow"
        ]
    }

One thing you can do is to enable the CloudWatchLog and check the log to see if there is any error message shown. You can also use AWS IoT Lifecycle event by subscribing to `$aws/events/#` in your AWS IoT Console to see if there is any connect/disconnect event captured when the issue happens. Related documentation can be found here:
http://docs.aws.amazon.com/iot/latest/developerguide/cloud-watch-logs.html
http://docs.aws.amazon.com/iot/latest/developerguide/life-cycle-events.html

You mentioned that you were able to use the same configuration and credentials to connect using AWS IoT Embedded C SDK. Can you also attach your Embedded C SDK configuration header file? What is the policy attached to your certificate? Does it come with any restrictions?

The AWS IoT Python SDK also supports MQTT over WebSocket. Another thing you can try is to use IAM credentials and see if you can get connected. More instructions can be found here in the Credentials section:
https://github.com/aws/aws-iot-device-sdk-python#id3

Please let us know how it goes.

Thanks,
Liusu

**Comment by kadin2048** [comment]
Hi Liusu, and thanks for the reply.  I actually solved the problem; in case anyone else runs into the same thing, here is some additional information....

I checked the lifecycle events (subscribing to `$aws/events/#` in the MQTT client) and nothing was showing up.

When terminating `basicPubSub.py`, I noticed that the line at which it was hanging (shown in the exception trace triggered by the KeyboardInterrupt) mentioned trying to get the device's hostname.  This led me to investigate the network stack.

As it turned out, although everything on the RPi was functioning properly, there was a mismatch in the hostname as defined in `/etc/hostname` and `/etc/hosts`, and the one being provided via DHCP.  This was seemingly causing a silent OpenSSL error (but, strangely, only to Python, not to the Embedded C SDK).  I suspect this has to do with how Python retrieves the hostname vs. the C implementation, but I'm not sure.

Anyway, the solution was to fix the hostname in `/etc/hostname` and `/etc/hosts`, **reboot** (it didn't work without rebooting, even though the hostname was being reported properly by `os.socket` in Python after updating the files!), and then basicPubSub.py started working.

Hopefully this will be of use to someone else in the future.
